src/ucs/memory/rcache_int.h

```diff
@@ -29,47 +29,48 @@ enum {
-    ucs_pgtable_t            pgtable;     /**< page table to hold the regions */
+    pthread_rwlock_t    pgt_lock;             /**< Protects the page table and all
+                                                   regions whose refcount is 0 */
+    ucs_pgtable_t       pgtable;              /**< page table to hold the regions */
 
 
-    ucs_spinlock_t           lock;        /**< Protects 'mp', 'inv_q' and 'gc_list'.
-                                               This is a separate lock because we
-                                               may want to invalidate regions
-                                               while the page table lock is held by
-                                               the calling context.
-                                               @note: This lock should always be
-                                               taken **after** 'pgt_lock'. */
-    ucs_mpool_t              mp;          /**< Memory pool to allocate entries for
-                                               inv_q and page table entries, since
-                                               we cannot use regular malloc().
-                                               The backing storage is original mmap()
-                                               which does not generate memory events */
-    ucs_queue_head_t         inv_q;       /**< Regions which were invalidated during
-                                               memory events */
-    ucs_list_link_t          gc_list;     /**< list for regions to destroy, regions
-                                               could not be destroyed from memhook */
+    ucs_spinlock_t      lock;                 /**< Protects 'mp', 'inv_q' and 'gc_list'.
+                                                   This is a separate lock because we
+                                                   may want to invalidate regions
+                                                   while the page table lock is held by
+                                                   the calling context.
+                                                   @note: This lock should always be
+                                                   taken **after** 'pgt_lock'. */
+    ucs_mpool_t         mp;                   /**< Memory pool to allocate entries for
+                                                   inv_q and page table entries, since
+                                                   we cannot use regular malloc().
+                                                   The backing storage is original mmap()
+                                                   which does not generate memory events */
+    ucs_queue_head_t    inv_q;                /**< Regions which were invalidated during
+                                                   memory events */
+    size_t              pending_cleanup_size; /**< Total size of the regions in gc_list and in inv_q */
```

Comment:
maybe rename it to "unreleased_size" and rename the parameter "cleanup_thresh" to "max_unreleased"?
(not sure it's a better name, but pending_cleanup_size / cleanup_thresh also sounds a bit weird)